File: WayforkWindows/app/lib/core/store/store_repository.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:wayfork/core/model/store.dart';

final class StoreLoadResult {
  const StoreLoadResult({required this.store, this.corruptBackup});

  final Store store;
  final File? corruptBackup;

  @override
  bool operator ==(Object other) =>
      other is StoreLoadResult &&
      store == other.store &&
      corruptBackup?.path == other.corruptBackup?.path;

  @override
  int get hashCode => Object.hash(store, corruptBackup?.path);
}

enum StoreRepositoryError { newerSchema }

final class StoreRepositoryException implements Exception {
  const StoreRepositoryException.newerSchema({
    required this.found,
    required this.supported,
  }) : kind = StoreRepositoryError.newerSchema;

  final StoreRepositoryError kind;
  final int found;
  final int supported;

  @override
  bool operator ==(Object other) =>
      other is StoreRepositoryException &&
      kind == other.kind &&
      found == other.found &&
      supported == other.supported;

  @override
  int get hashCode => Object.hash(kind, found, supported);

  @override
  String toString() =>
      'Store schema $found is newer than supported schema $supported';
}

/// Loads and saves `store.json` with debounced, atomic writes.
final class StoreRepository {
  StoreRepository(
    this.directory, {
    this.debounce = const Duration(milliseconds: 300),
  });

  static const fileName = 'store.json';

  final Directory directory;
  final Duration debounce;
  Store? _pending;
  Timer? _timer;

  File get file => File(_join(directory.path, fileName));

  bool get hasPendingChanges => _pending != null;

  static Directory defaultDirectory() {
    final environment = Platform.environment;
    if (Platform.isWindows) {
      final localAppData = environment['LOCALAPPDATA'];
      if (localAppData == null || localAppData.isEmpty) {
        throw StateError('LOCALAPPDATA is not set');
      }
      return Directory('$localAppData\\Wayfork');
    }
    final home = environment['HOME'];
    if (home == null || home.isEmpty) throw StateError('HOME is not set');
    return Platform.isMacOS
        ? Directory('$home/Library/Application Support/Wayfork')
        : Directory('$home/.local/share/wayfork');
  }

  Future<StoreLoadResult> load() async {
    if (!await file.exists()) return StoreLoadResult(store: Store.empty);
    final bytes = await file.readAsBytes();
    try {
      return StoreLoadResult(store: StoreCodec.decode(utf8.decode(bytes)));
    } on StoreCodecException catch (error) {
      if (error.kind == StoreCodecError.newerSchema) {
        throw StoreRepositoryException.newerSchema(
          found: error.found!,
          supported: error.supported!,
        );
      }
      return _moveCorruptAside();
    } on Object {
      return _moveCorruptAside();
    }
  }

  void save(Store store) {
    _pending = store;
    _timer?.cancel();
    _timer = Timer(debounce, () {
      unawaited(flush().catchError((Object _) {}));
    });
  }

  Future<void> flush() async {
    _timer?.cancel();
    _timer = null;
    final store = _pending;
    if (store == null) return;
    await write(store);
  }

  Future<void> write(Store store) async {
    await directory.create(recursive: true);
    final temporary = File('${file.path}.tmp');
    if (await temporary.exists()) await temporary.delete();
    await temporary.writeAsString(StoreCodec.encode(store), flush: true);
    await temporary.rename(file.path);
    if (_pending == store) _pending = null;
  }

  Future<StoreLoadResult> _moveCorruptAside() async {
    final backup = File('${file.path}.corrupt-${_timestamp(DateTime.now())}');
    if (await backup.exists()) await backup.delete();
    await file.rename(backup.path);
    return StoreLoadResult(store: Store.empty, corruptBackup: backup);
  }

  static String _timestamp(DateTime date) {
    final utc = date.toUtc();
    String two(int value) => value.toString().padLeft(2, '0');
    String four(int value) => value.toString().padLeft(4, '0');
    return '${four(utc.year)}${two(utc.month)}${two(utc.day)}-'
        '${two(utc.hour)}${two(utc.minute)}${two(utc.second)}';
  }
}

String _join(String directory, String name) {
  final separator = Platform.isWindows ? '\\' : '/';
  return directory.endsWith(separator)
      ? '$directory$name'
      : '$directory$separator$name';
}

File: macos/WayforkCore/Sources/WayforkCore/Store/StoreRepository.swift
import Foundation

/// Loads and saves `store.json` (docs/design/01-data-model.md, "Persistence"): atomic
/// writes debounced after the last change, corrupt files moved aside, newer schemas refused.
public actor StoreRepository {
    public static let fileName = "store.json"

    public struct LoadResult: Sendable, Equatable {
        public var store: Store
        /// Set when the previous file was unreadable and has been renamed to this path
        /// (`store.corrupt` in the UI).
        public var corruptBackup: URL?
    }

    public enum Error: Swift.Error {
        case newerSchema(found: Int, supported: Int)
    }

    public let directory: URL
    public var fileURL: URL { directory.appendingPathComponent(StoreRepository.fileName) }

    private let debounce: Duration
    private var pending: Store?
    private var flushTask: Task<Void, Never>?
    private let fileManager = FileManager.default

    /// - Parameter directory: `~/Library/Application Support/Wayfork` in the app.
    public init(directory: URL, debounce: Duration = .milliseconds(300)) {
        self.directory = directory
        self.debounce = debounce
    }

    public static func defaultDirectory() -> URL {
        let base =
            FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent(
                "Library/Application Support")
        return base.appendingPathComponent("Wayfork", isDirectory: true)
    }

    /// Missing file → empty store. Corrupt file → renamed to `store.json.corrupt-<timestamp>`,
    /// empty store plus `corruptBackup`. Newer schema → `Error.newerSchema` (never touched).
    public func load() throws -> LoadResult {
        guard fileManager.fileExists(atPath: fileURL.path) else {
            return LoadResult(store: .empty, corruptBackup: nil)
        }
        let data = try Data(contentsOf: fileURL)
        do {
            return LoadResult(store: try StoreCodec.decode(data), corruptBackup: nil)
        } catch StoreCodec.Error.newerSchema(let found, let supported) {
            throw Error.newerSchema(found: found, supported: supported)
        } catch {
            let backup = directory.appendingPathComponent(
                "\(StoreRepository.fileName).corrupt-\(Self.timestamp())")
            try? fileManager.removeItem(at: backup)
            try fileManager.moveItem(at: fileURL, to: backup)
            return LoadResult(store: .empty, corruptBackup: backup)
        }
    }

    /// Schedules a write; consecutive calls within the debounce window collapse into one.
    public func save(_ store: Store) {
        pending = store
        flushTask?.cancel()
        let delay = debounce
        flushTask = Task { [weak self] in
            try? await Task.sleep(for: delay)
            guard !Task.isCancelled else { return }
            try? await self?.flush()
        }
    }

    /// Writes the pending store now, if any.
    public func flush() throws {
        flushTask?.cancel()
        flushTask = nil
        guard let store = pending else { return }
        try write(store)
        pending = nil
    }

    /// Writes immediately, bypassing the debounce.
    public func write(_ store: Store) throws {
        try fileManager.createDirectory(
            at: directory, withIntermediateDirectories: true,
            attributes: [.posixPermissions: 0o700])
        let data = try StoreCodec.encode(store)
        try data.write(to: fileURL, options: .atomic)
        try fileManager.setAttributes([.posixPermissions: 0o600], ofItemAtPath: fileURL.path)
        pending = nil
    }

    public var hasPendingChanges: Bool { pending != nil }

    private static func timestamp() -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = TimeZone(identifier: "UTC")
        formatter.dateFormat = "yyyyMMdd-HHmmss"
        return formatter.string(from: Date())
    }
}
